### PR TITLE
[move-prover] support remapping of local variables during spec translation

### PR DESCRIPTION
**Pull request by matbd**
## Motivation

This PR
* simplifies the translation of quantified formulas in the case of vector and primitive ranges.
* unblocks the generation of triggers in those cases (see https://github.com/diem/diem/pull/7501).

Specifically, `forall a in V  : exp ` used to translate to:
```
(var $range_0 := $Range($Integer(0), $vlen_value(V));
 $Boolean((
   forall $i_1: int :: $InRange($range_0, $i_1) ==> (var a := $Integer($i_1); EXP[a])
)))
```
Note that a trigger would be added after `::` and before `var a := ..` is defined which is problematic.

After this PR, we do not quantify over a temporary but rather remap the variable into a Boogie expression:
```
(var $range_0 := $Range($Integer(0), $vlen_value(V));
 $Boolean((
   forall a: int :: $InRange($range_0, a) ==> EXP[a := "$Integer(a)"]
)))
```

## Test Plan

CI

### CODE REVIEW

File: language/move-prover/boogie-backend/src/spec_translator.rs [hunk 1]
```diff
@@ -47,6 +47,8 @@ pub struct SpecTranslator<'env> {
     /// Set of items which have been already traced. This is used to avoid redundant tracing
     /// of expressions whose value has been already tracked.
     traced_items: RefCell<BTreeSet<TraceItem>>,
+    /// Local variables with a known translation.
+    assigned_local_vars: RefCell<HashMap<Symbol, Vec<String>>>,
```
Comment (wrwg) [inline language/move-prover/boogie-backend/src/spec_translator.rs:51]: We uniformly use BTreeX collections in the code base (of Diem even I think) because of the potential non-determinism of HashMap. A map which is never enumerated like this one should be save. But as far as performance goes, its also interesting whether a map which has 2-3 elements in average might be faster as BTree.

File: language/move-prover/boogie-backend/src/spec_translator.rs [hunk 2]
```diff
@@ -435,6 +463,12 @@ impl<'env> SpecTranslator<'env> {
     }
 
     fn translate_local_var(&self, node_id: NodeId, name: Symbol) {
+        if let Some(values) = self.assigned_local_vars.borrow().get(&name) {
```
Comment (wrwg) [inline language/move-prover/boogie-backend/src/spec_translator.rs:466]: Don't we need to deal with shadowing? Or is the context how we use this feature (in quantifiers) makes this unnecessary?

File: language/move-prover/boogie-backend/src/spec_translator.rs [hunk 3]
```diff
@@ -729,24 +763,31 @@ impl<'env> SpecTranslator<'env> {
     ) {
         let loc = self.env.get_node_loc(node_id);
         // Translate range expressions.
-        let mut range_tmps = HashMap::new();
+        let mut vec_range_tmps = HashMap::new();
```
Comment (wrwg) [inline language/move-prover/boogie-backend/src/spec_translator.rs:766]: Wondering whether it is time to break this  function down by introducing a few helper functions, or specializing for the multiple cases.

File: language/move-prover/src/spec_translator.rs [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 use std::{
     cell::RefCell,
-    collections::{BTreeSet, HashMap},
+    collections::{hash_map, BTreeSet, HashMap},
```
Comment (wrwg) [inline language/move-prover/src/spec_translator.rs:8]: Ugh. Ugly we have to make this change twice. Perhaps better we add a test  `quantifiers.move` to tests/sources/functional and only do this, moving forward, for --v2. Like a number of other tests, if the test contains the line `flag: --v2` it will be run for v2 compilation scheme. We should be days away from having moved everything to v2. A new `quantifiers` test would anyway be good to have to test this for corner cases (e.g. including shadowing).

**Comment by wrwg** [comment]
> Hey hey!
> 
> It looks like #wannabe-land-blocking found some issues that are not yet land blocking, but will be in the future.
> Please fix these:
> 
> click here to reveal them
> ```
> warning: integer arithmetic detected
>    --> language/move-prover/boogie-backend/src/spec_translator.rs:882:27
>     |
> 882 |                     .take(3 + vec_range_tmps.len() + prim_range_tmps.len())
>     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#integer_arithmetic
> warning: integer arithmetic detected
> --> language/move-prover/src/spec_translator.rs:2347:27
> |
> 2347 |                     .take(3 + vec_range_tmps.len() + prim_range_tmps.len())
> |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> |
> = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#integer_arithmetic
> ```

Please also fix this if you come along in code you touch (using usize::checked_add or something). I have some github issues assigned to me by Bob to fix this in the prover as it will soon be land blocking.

File: language/move-prover/boogie-backend/src/spec_translator.rs (hunk 2 above)
Comment (matbd) [reply language/move-prover/boogie-backend/src/spec_translator.rs:466]: If assignments are stacked (e.g. in nested quantifiers) and symbols are clashing, it will (correctly) use the inner-most assignment. However, assigned Boogie expressions are final: they cannot be further remapped. This does indeed limit the Boogie variables that one can safely capture in those expressions -- I would say -- to the original variable but fresh temporaries.

File: language/move-prover/boogie-backend/src/spec_translator.rs (hunk 1 above)
Comment (matbd) [reply language/move-prover/boogie-backend/src/spec_translator.rs:51]: Sure I can switch to BTreeMap

File: language/move-prover/boogie-backend/src/spec_translator.rs (hunk 2 above)
Comment (wrwg) [reply language/move-prover/boogie-backend/src/spec_translator.rs:466]: What's about `forall x in v: P[{let x = E[x]; E'[x]}]`?

Comment (matbd) [reply language/move-prover/boogie-backend/src/spec_translator.rs:466]: I'm not sure to follow. The replacement is supposed to be a stringified Boogie expression.

File: language/move-prover/boogie-backend/src/spec_translator.rs (hunk 3 above)
Comment (matbd) [reply language/move-prover/boogie-backend/src/spec_translator.rs:766]: TBH I haven't found a non-artificial way to break down this function.

File: language/move-prover/src/spec_translator.rs (hunk 4 above)
Comment (matbd) [reply language/move-prover/src/spec_translator.rs:8]: Fair point. Let me add a test

File: language/move-prover/boogie-backend/src/spec_translator.rs (hunk 2 above)
Comment (wrwg) [reply language/move-prover/boogie-backend/src/spec_translator.rs:466]: I mean you can have the variable introduced by a quantifier be shadowed by a variable of the same name introduced in a let anywhere in the expressions used in a quantifier.

The translate_local function will map the local name `x` to the boogie expression, even thought it might be shadowed and mean something else in E'[x] above. Not 100% sure, but at least that is a concern I have.

Comment (matbd) [reply language/move-prover/boogie-backend/src/spec_translator.rs:466]: I'm going to add a test file to support the discussion.

Comment (matbd) [reply language/move-prover/boogie-backend/src/spec_translator.rs:466]: Ok so we do have any issue with shadowed variables.
Currently, we generate Boogie variables that are shadowing themselves in the same manner as the original code.

File: language/move-prover/boogie-backend/src/spec_translator.rs (hunk 3 above)
Comment (wrwg) [reply language/move-prover/boogie-backend/src/spec_translator.rs:766]: SG. It was more a generic review comment.
